OrbitVulkanLayer/DeviceManager.h

```diff
@@ -0,0 +1,99 @@
+ * `TrackLogicalDevice` establishes that mapping, while `UntrackLogicalDevice` can be used to
+ * release that mapping. To retrieve a logical device's physical device, use `GetPhysicalDevice`.
+ *
+ * For each physical device, it also provides the `VkPhysicalDeviceProperties`, which can be queried
+ * using `GetPhysicalDeviceProperties`. The `DeviceManager` is responsible for retrieving this
+ * information (using `vkGetPhysicalDeviceProperties`). These properties can be used e.g. for
+ * converting clock cycles to nanosecond timestamps.
+ *
+ * Thread-Safety: This class is internally synchronized (using read/write locks) and can be safely
+ * accessed from different threads.
+ */
+template <class DispatchTable>
+class DeviceManager {
+ public:
+  explicit DeviceManager(DispatchTable* dispatch_table) : dispatch_table_(dispatch_table) {}
+
+  void TrackLogicalDevice(VkPhysicalDevice physical_device, VkDevice device) {
+    absl::WriterMutexLock lock(&mutex_);
+    CHECK(!logica_device_to_physical_device_.contains(device));
+    logica_device_to_physical_device_[device] = physical_device;
+
+    CHECK(!physical_device_to_logical_devices_.contains(physical_device) ||
+          !physical_device_to_logical_devices_.at(physical_device).contains(device));
+    if (!physical_device_to_logical_devices_.contains(physical_device)) {
+      physical_device_to_logical_devices_[physical_device] = {device};
+    } else {
+      physical_device_to_logical_devices_.at(physical_device).insert(device);
+    }
+
+    if (physical_device_to_properties_.contains(physical_device)) {
+      return;
+    }
+
+    VkPhysicalDeviceProperties properties;
+    dispatch_table_->GetPhysicalDeviceProperties(physical_device)(physical_device, &properties);
+    physical_device_to_properties_[physical_device] = properties;
+  }
+
+  [[nodiscard]] VkPhysicalDevice GetPhysicalDeviceOfLogicalDevice(VkDevice device) {
+    absl::ReaderMutexLock lock(&mutex_);
+    CHECK(logica_device_to_physical_device_.contains(device));
+    return logica_device_to_physical_device_.at(device);
+  }
+
+  void UntrackLogicalDevice(VkDevice device) {
+    absl::WriterMutexLock lock(&mutex_);
+    CHECK(logica_device_to_physical_device_.contains(device));
+    VkPhysicalDevice physical_device = logica_device_to_physical_device_.at(device);
+    logica_device_to_physical_device_.erase(device);
+
+    CHECK(physical_device_to_logical_devices_.contains(physical_device));
+    absl::flat_hash_set<VkDevice>& logical_devices =
+        physical_device_to_logical_devices_.at(physical_device);
+    CHECK(logical_devices.contains(device));
+    logical_devices.erase(device);
+
+    if (!logical_devices.empty()) {
+      return;
+    }
+    physical_device_to_logical_devices_.erase(physical_device);
+    physical_device_to_properties_.erase(logica_device_to_physical_device_.at(device));
+  }
+
+  [[nodiscard]] VkPhysicalDeviceProperties GetPhysicalDeviceProperties(VkPhysicalDevice device) {
```

Comment:
Should the argument be `physical_device` for general consistency?

Reply:
Agree, thanks!